core/src/main/java/org/jruby/ir/IRMethod.java

```diff
@@ -138,17 +139,33 @@ public InterpreterContext builtInterpreterContext() {
      *
      * @return appropriate interpretercontext
      */
-    public synchronized InterpreterContext builtInterperterContextForJavaConstructor() {
-        InterpreterContext interpreterContext = builtInterpreterContext();
+    public ExitableInterpreterContext builtInterperterContextForJavaConstructor() {
+        ExitableInterpreterContext interpreterContextForJavaConstructor = this.interpreterContextForJavaConstructor;
+        if (interpreterContextForJavaConstructor == null) {
+            synchronized (this) {
+                interpreterContextForJavaConstructor = this.interpreterContextForJavaConstructor;
+                if (interpreterContextForJavaConstructor == null) {
+                    interpreterContextForJavaConstructor = doBuiltInterperterContextForJavaConstructor();
+                    this.interpreterContextForJavaConstructor = interpreterContextForJavaConstructor;
+                }
+            }
+        }
+        return interpreterContextForJavaConstructor == ExitableInterpreterContext.NULL ? null : interpreterContextForJavaConstructor;
+    }
+
+    private ExitableInterpreterContext interpreterContextForJavaConstructor; // not volatile - worst case we built it more than once
```

Comment:
I think this _should_ be volatile, since (Exitable)InterpreterContext is not fully immutable (i.e. not all of its fields are final and contain only primitive values or other full immutable objects).

AIUI, using double-checked locking in Java without volatile risks other threads possibly accessing the object in a partially constructed state (i.e. the JVM is allowed to inline the constructor call and reorder the assignment of its return value to happen before the object is fully constructed). Using volatile will prevent this reordering.

(So would making all fields of the constructed object final, although the construction of any objects they reference could still be subject to reordering. Here, I see you've made the superCall and exitIPC fields of ExitableInterpreterContext final below, but the parent InterpreterContext class still seems to have non-final fields. In any case, even if we _could_ guarantee thread safety in this manner, I'd still much rather pay the fairly minor cost of volatile just to be sure.)

Reply:
yeah in theory I guess it makes sense but I also wonder if that is ever the case.

Reply:
yeah in theory I guess it makes sense but I also wonder if that is ever the case.

this is pretty much modeled after the (inherited) `protected InterpreterContext interpreterContext` which doesn't even [double check](https://github.com/jruby/jruby/blob/9.3.10.0/core/src/main/java/org/jruby/ir/IRMethod.java#L192-L198), upon lazy building, and relies upon the builder setting the (non-volatile) `interpreterContext` in a [synchronized method](https://github.com/jruby/jruby/blob/9.3.10.0/core/src/main/java/org/jruby/ir/IRMethod.java#L203). so if the re-ordering was an actual issue happening in practice we would have seen it with the `interpreterContext` (has been like this for years).

thing we could redesign the hierarchy a bit but was trying to keep this minimal (at least for 9.3) ...

Reply:
The IRMethod::hasBeenBuilt‎ method checks if `defNode == null`, where `defNode` is volatile and is set to null at the end of IRMethod::buildMethodImpl‎. I'm not 100% sure, but I _think_ that's enough under the Java memory model to prevent the problematic reordering. It still feels a bit sketchy (as in, someone could easily break it and introduce a very subtle concurrency bug without noticing), but I think it's technically safe. We don't have anything like that here, though.

Reply:
I took another look at this after the changes in bef9cce, but this still doesn't look 100% safe to me without volatile, since InterpreterContext still has non-final fields that should be initialized in the constructor for correct operation.

In particular, I believe it would be permissible for the JVM to reorder the `setInstructions` call in the InterpreterContext constructor to happen after the assignment to `interpreterContextForJavaConstructor`. If another thread then happened to grab the partially constructed object and call `getInstructions` or `getEngine` on it while `instructions` was still null, bad things could happen.

And this isn't something that can be easily fixed AFAICT, since deferring the call to `setInstructions` is apparently a valid use case (when using the other constructor that doesn't call it).

Honestly, I'd just bite the bullet and make this field volatile. Or at least benchmark it with and without volatile and see if there's even any measurable performance difference. At least, even with volatile, it should still be faster than it was before this PR.

Reply:
> I took another look at this after the changes in https://github.com/jruby/jruby/commit/bef9cce3577ec1ba8577128a62f11ef7d6999e68,

that one is just a low hanging fruit - making it all immutable is not possible due the lazy design of the instructions as you already noted.

> The IRMethod::hasBeenBuilt‎ method checks if `defNode == null`, where `defNode` is volatile

thanks I might have missed the volatile part there. will take another look at this, if we can do smt minimal, and if not probably just set the `volatile` as suggested ... given the normal `InterpreterContext` is behind a volatile check.

Reply:
updated [original commit](https://github.com/jruby/jruby/pull/7762/commits/d0710cc26875ff52c45e09ac2d5d9da73036ab90#diff-2d9019354e47c1489773d64b60d53afe974ef7676db097c83b23e46ca2fe70f3R156) so it's easy to cherry-pick and patch